snapshot/driver.go

```diff
@@ -0,0 +1,310 @@
+//
+// A snapshot represents a filesystem state. Every snapshot has a parent, where
+// the empty parent is represented by the empty string.
+//
+// For convention, we define the following terms to be used throughout this
+// interface for driver implementations:
+//
+// 	`name` - refers to a forkable snapshot, typically read only
+// 	`key` - refers to an active transaction, either a prepare or view
+// 	`parent` - refers to the parent in relation to a name
+//
+// TODO(stevvooe): Update this description when things settle.
+//
+// Importing a Layer
+//
+// To import a layer, we simply have the Manager provide a list of
+// mounts to be applied such that our dst will capture a changeset. We start
+// out by getting a path to the layer tar file and creating a temp location to
+// unpack it to:
+//
+//	layerPath, tmpLocation := getLayerPath(), mkTmpDir() // just a path to layer tar file.
+//
+// We then use a Manager to prepare the temporary location as a
+// snapshot point:
+//
+// 	sm := NewManager()
+//	mounts, err := sm.Prepare(tmpLocation, "")
+// 	if err != nil { ... }
+//
+// Note that we provide "" as the parent, since we are applying the diff to an
+// empty directory. We get back a list of mounts from Manager.Prepare.
+// Before proceeding, we perform all these mounts:
+//
+//	if err := MountAll(mounts); err != nil { ... }
+//
+// Once the mounts are performed, our temporary location is ready to capture
+// a diff. In practice, this works similar to a filesystem transaction. The
+// next step is to unpack the layer. We have a special function unpackLayer
+// that applies the contents of the layer to target location and calculates the
+// DiffID of the unpacked layer (this is a requirement for docker
+// implementation):
+//
+//	layer, err := os.Open(layerPath)
+//	if err != nil { ... }
+// 	digest, err := unpackLayer(tmpLocation, layer) // unpack into layer location
+// 	if err != nil { ... }
+//
+// When the above completes, we should have a filesystem the represents the
+// contents of the layer. Careful implementations should verify that digest
+// matches the expected DiffID. When completed, we unmount the mounts:
+//
+//	unmount(mounts) // optional, for now
+//
+// Now that we've verified and unpacked our layer, we create a location to
+// commit the actual diff. For this example, we are just going to use the layer
+// digest, but in practice, this will probably be the ChainID:
+//
+// 	diffPath := filepath.Join("/layers", digest) // name location for the uncompressed layer digest
+//	if err := sm.Commit(diffPath, tmpLocation); err != nil { ... }
+//
+// Now, we have a layer in the Manager that can be accessed with the
+// opaque diffPath provided during commit.
+//
+// Importing the Next Layer
+//
+// Making a layer depend on the above is identical to the process described
+// above except that the parent is provided as diffPath when calling
+// Manager.Prepare:
+//
+// 	mounts, err := sm.Prepare(tmpLocation, parentDiffPath)
```

Comment:
`s/parentDiffPath/diffPath` to make it easier for reader to make the connection?